src/treetensornetworks/abstracttreetensornetwork.jl

```diff
@@ -365,20 +365,14 @@ end
 # Inner products
 #
 
-# TODO: implement using multi-graph disjoint union
-function inner(
-  y::AbstractTTN, A::AbstractTTN, x::AbstractTTN; root_vertex=default_root_vertex(x, A, y)
-)
-  traversal_order = reverse(post_order_dfs_vertices(x, root_vertex))
-  check_hascommoninds(siteinds, A, x)
-  check_hascommoninds(siteinds, A, y)
+# TODO: Remove dispatch on AbstractTTN (unless we want to trigger warning for trees if BP is nonexact)  
+function inner(y::AbstractTTN, A::AbstractTTN, x::AbstractTTN; kwargs...)
   ydag = sim(dag(y); sites=[])
-  x = sim(x; sites=[])
-  O = ydag[root_vertex] * A[root_vertex] * x[root_vertex]
-  for v in traversal_order[2:end]
-    O = O * ydag[v] * A[v] * x[v]
-  end
-  return O[]
+  blf = BilinearFormNetwork(A, ydag, x)
+  blf_scalar_bp = contract_with_BP(
```

Comment:
If we just use `contract` here, then we can allow an `alg` keyword argument to get passed from `inner`, and users can control what algorithm is used to contract the network.

To handle partitioning in the more general case, something we could do is that instead of passing an `ITensorNetwork` and a partitioning, we pass a `PartitionedGraph` to `contract`, then it is up to the `contract` backends to contract the partitioned tensor network (either using or ignoring the partitioning).